### PR TITLE
Search rescoring modes

### PR DESCRIPTION
**Pull request by IvanPleshkov**
Issue https://github.com/qdrant/qdrant/issues/2624

For binary quantization, we want to rescore search results by default. Because without rescoring plain search result will be missed, the reason is in large difference between original and quantized scores.
On the other hand, we want to disable rescoring by default. Because of performance reasons.

As a possible solution, this PR introduces an `Auto` rescoring mode which enables rescoring for binary quantization and disables for another kind.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: docs/redoc/master/openapi.json [hunk 1]
```diff
@@ -6191,9 +6191,7 @@
             "type": "boolean"
           },
           "rescore": {
-            "description": "If true, use original vectors to re-score top-k results. Might require more time in case if original vectors are stored on disk. Default is false.",
-            "default": false,
-            "type": "boolean"
+            "$ref": "#/components/schemas/Rescoring"
```
Comment (generall) [inline docs/redoc/master/openapi.json:6194]: this is not backward compatible.

Comment (IvanPleshkov) [reply docs/redoc/master/openapi.json:6194]: okay, will revert to `Option<bool>` and will add more text comments about `None` behaviuor

**Review by generall — CHANGES_REQUESTED** [review]
It should be backward compatible

**Comment by IvanPleshkov** [comment]
> It should be backward compatible

fixed

File: lib/segment/src/types.rs [hunk 2]
```diff
@@ -284,9 +284,9 @@ pub struct QuantizationSearchParams {
 
     /// If true, use original vectors to re-score top-k results.
     /// Might require more time in case if original vectors are stored on disk.
-    /// Default is false.
-    #[serde(default = "default_quantization_rescore_value")]
-    pub rescore: bool,
+    /// If not set, qdrant decides automatically apply rescoring or not.
```
Comment (coszio) [inline lib/segment/src/types.rs:287]:
```suggestion
    ///
    /// If not set, qdrant decides automatically to re-score or not, depending on quantization settings.
```

Should we include the example of binary quantization having a side-effect on score?

**Review by coszio — COMMENTED** [review]
Looks good, only comments about doc wording

**Review by generall — APPROVED** [review]

**Pull request merged**